### PR TITLE
rgw: back off bucket sync on failures, don't store marker

### PR DESCRIPTION
**Pull request by yehudasa**
Fixes: http://tracker.ceph.com/issues/16742

If we fail on any single entry in bucket, skip updating the marker tracker
so that next time we'll go over that entry, and back off. This will trigger
a report to the data sync error repo and eventually a retry on the failing
object.

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -2286,33 +2288,41 @@ int RGWBucketShardFullSyncCR::operate()
           while (collect(&ret, lease_stack)) {
             if (ret < 0) {
               ldout(sync_env->cct, 0) << "ERROR: a sync operation returned error" << dendl;
+              sync_status = ret;
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc]: `collect()` never returns true, so this block won't execute. removing the `while()` should work. same with `RGWBucketShardIncrementalSyncCR` below

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -2512,33 +2541,55 @@ int RGWBucketShardIncrementalSyncCR::operate()
...
+      while (again) {
+        again = collect(&ret, lease_stack);
+        if (ret < 0) {
+          ldout(sync_env->cct, 0) << "ERROR: a sync operation returned error" << dendl;
+          sync_status = ret;
+          /* we have reported this error */
+        }
+        /* not waiting for child here */
+      }
+    }
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc]: looks correct with the new `collect()`. this logic is almost identical to `RGWCoroutine::drain_children()` though, maybe we could find a way to get the error codes out of that instead?

File: src/rgw/rgw_sync.h [hunk 3]
```diff
@@ -363,13 +355,23 @@ class RGWSyncShardMarkerTrack {
...
-    last_stored_marker = high_marker;
-    return store_marker(high_marker, high_entry.pos, high_entry.timestamp);
+
+    auto last = i;
+    --i;
+    const T& high_marker = i->first;
+    marker_entry& high_entry = i->second;
+    RGWCoroutine *cr = store_marker(high_marker, high_entry.pos, high_entry.timestamp);
+    finish_markers.erase(finish_markers.begin(), last);
+    return cr;
```
Comment (cbodley) [inline src/rgw/rgw_sync.h]: looks good. it does look like a fix for the meta sync issue in bz 1357641

File: src/rgw/rgw_data_sync.cc (hunk 2 above)
Comment (yehudasa) [inline src/rgw/rgw_data_sync.cc]: @cbodley yes, this needs to be cleaned up. But we should probably do it outside the scope of this bug.

**Pull request merged**